### PR TITLE
Declaration transform

### PR DESCRIPTION
**Pull request by filipesilva**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
```
fix(bazel): also pass afterDeclarations transformers to emitWithTsickle

    Tsickle supports afterDeclarations transformers, but these were not being passed in.

refactor(compiler-cli): refactor import adding logic into helper

    This logic will be common to transforms that add imports. Using it as a helper helps reduce duplication

refactor(compiler-cli): use a transformer for dts files

    The current DtsFileTransformer works by intercepting file writes and editing the source string directly.

    This PR refactors it as a afterDeclaration transform in order to fit better in the TypeScript API.

    This is part of a greater effort of converting ngtsc to be usable as a TS transform plugin.

refactor(compiler-cli): return TS nodes from TypeTranslatorVisitor

    The TypeTranslatorVisitor visitor returned strings because before it wasn't possible to transform declaration files directly through the TypeScript custom transformer API.

    Now that's possible though, so it should return nodes instead.
```
## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/translator/src/translator.ts [hunk 1]
```diff
@@ -294,44 +294,44 @@ class ExpressionTranslatorVisitor implements ExpressionVisitor, StatementVisitor
...
 
-  visitExpressionType(type: ExpressionType, context: Context): string {
-    const exprStr = type.value.visitExpression(this, context);
-    if (type.typeParams !== null) {
-      const typeSegments = type.typeParams.map(param => param.visitType(this, context));
-      return `${exprStr}<${typeSegments.join(', ')}>`;
-    } else {
-      return exprStr;
-    }
+  visitExpressionType(type: ExpressionType, context: Context): ts.TypeReferenceType {
```
Comment (filipesilva) [inline packages/compiler-cli/src/ngtsc/translator/src/translator.ts]: Expressions here seem to actually be `ts.TypeReferenceType`.

File: packages/compiler-cli/src/ngtsc/translator/src/translator.ts [hunk 2]
```diff
@@ -365,28 +365,27 @@ export class TypeTranslatorVisitor implements ExpressionVisitor, TypeVisitor {
...
-      return `'${escaped}'`;
-    } else {
-      return `${ast.value}`;
-    }
+  visitLiteralExpr(ast: LiteralExpr, context: Context): ts.LiteralExpression {
+    return ts.createLiteral(ast.value as string);
   }
 
-  visitExternalExpr(ast: ExternalExpr, context: Context): string {
+  visitExternalExpr(ast: ExternalExpr, context: Context): ts.TypeReferenceType {
```
Comment (filipesilva) [inline packages/compiler-cli/src/ngtsc/translator/src/translator.ts]: Expressions here seem to actually be `ts.TypeReferenceType`.

File: packages/compiler-cli/src/ngtsc/translator/src/translator.ts [hunk 3]
```diff
@@ -417,37 +416,48 @@ export class TypeTranslatorVisitor implements ExpressionVisitor, TypeVisitor {
...
+      return node;
     } else {
       throw new Error(
           `Unsupported WrappedNodeExpr in TypeTranslatorVisitor: ${ts.SyntaxKind[node.kind]}`);
     }
   }
 
-  visitTypeofExpr(ast: TypeofExpr, context: Context): string {
-    return `typeof ${ast.expr.visitExpression(this, context)}`;
+  visitTypeofExpr(ast: TypeofExpr, context: Context): ts.TypeOfExpression {
```
Comment (filipesilva) [inline packages/compiler-cli/src/ngtsc/translator/src/translator.ts]: TypeOfs here seem to actually be `ts.TypeQuery` instead.

File: packages/compiler-cli/src/ngtsc/program.ts [hunk 4]
```diff
@@ -232,6 +228,7 @@ export class NgtscProgram implements api.Program {
     const customTransforms = opts && opts.customTransformers;
     const beforeTransforms =
         [ivyTransformFactory(this.compilation !, this.reflector, this.importRewriter, this.isCore)];
+    const afterDeclarationsTransforms = [declarationTransformFactory(this.compilation !)];
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/program.ts]: If you change line 216 above to be:

```ts
const compilation = this.ensureAnalyzed();
```

then I think you could use that constant here and avoid the `!` operator.

File: packages/compiler-cli/src/ngtsc/program.ts [hunk 5]
```diff
@@ -253,6 +250,7 @@ export class NgtscProgram implements api.Program {
       customTransformers: {
         before: beforeTransforms,
         after: customTransforms && customTransforms.afterTs,
+        afterDeclarations: afterDeclarationsTransforms,
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/program.ts]: Nice!

**Review by petebacondarwin — APPROVED** [review]
LGTM! Nice refactoring to avoid string manipulation.

File: packages/compiler-cli/src/ngtsc/program.ts (hunk 4 above)
Comment (filipesilva) [reply packages/compiler-cli/src/ngtsc/program.ts]: SGTM, changed 👍

Comment (filipesilva) [reply packages/compiler-cli/src/ngtsc/program.ts]: SGTM, just changed it 👍

File: packages/compiler-cli/src/ngtsc/program.ts (hunk 5 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/program.ts]: 🎉

File: packages/compiler-cli/src/ngtsc/transform/src/declaration.ts [hunk 6]
```diff
@@ -32,36 +47,34 @@ export class DtsFileTransformer {
...
+  transform(file: ts.SourceFile, context: ts.TransformationContext): ts.SourceFile {
+    const visitor: ts.Visitor = (node: ts.Node): ts.VisitResult<ts.Node> => {
+      // This class declaration needs to have fields added to it.
+      if (ts.isClassDeclaration(node) && node.name !== undefined &&
+          this.ivyFields.has(node.name.text)) {
+        const decls = this.ivyFields.get(node.name.text) !;
+        const newMembers = decls.map(decl => {
+          const modifiers = [ts.createModifier(ts.SyntaxKind.StaticKeyword)];
+          const type = translateType(decl.type, this.imports);
+          const typeRef = ts.createTypeReferenceNode(ts.createIdentifier(type), undefined);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/transform/src/declaration.ts]: You, sir, are very clever.

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/transform/src/declaration.ts (hunk 6 above)
Comment (filipesilva) [reply packages/compiler-cli/src/ngtsc/transform/src/declaration.ts]: 😊

**Comment by alexeagle** [comment]
CARETAKER: ngbot status seems stuck. not kicking it so that @ocombe has a chance to investigate

**Pull request closed without merging**